packages/liveblocks-core/src/lib/objectToQuery.ts

```diff
@@ -0,0 +1,222 @@
+    isValueWithOperator(value)
+  ) as [string, Record<"startsWith", string>][];
+
+  filterList = filterList.concat(
+    getFiltersFromKeyValuePairsWithOperator(keyValuePairsWithOperator)
+  );
+
+  /**
+   * Case 3: Indexed key value pairs
+   * @example
+   * ```ts
+   * { key: { nestedKey: "foo" } }
+   * ```
+   * @example
+   * ```ts
+   * { key: { nestedKey: { startsWith: "fo" } } }
+   * ```
+   */
+  const indexedKeys = entries.filter(([_, value]) => {
+    return typeof value === "object" && !("startsWith" in value);
+  });
+
+  indexedKeys.forEach(([key, value]) => {
+    const nestedEntries = Object.entries(value);
+    const nestedKeyValuePairs = nestedEntries
+      .filter(([_, value]) => isSimpleValue(value))
+      .map(([nestedKey, nestedValue]) => [
+        formatFilterKey(key, nestedKey),
+        nestedValue,
+      ]) as [string, string | number | boolean][];
+
+    filterList = filterList.concat(
+      getFiltersFromKeyValuePairs(nestedKeyValuePairs)
+    );
+
+    const nestedKeyValuePairsWithOperator = nestedEntries
+      .filter(([_, value]) => isValueWithOperator(value))
+      .map(([nestedKey, nestedValue]) => [
+        formatFilterKey(key, nestedKey),
+        nestedValue,
+      ]) as [string, { startsWith: string }][];
+
+    filterList = filterList.concat(
+      getFiltersFromKeyValuePairsWithOperator(nestedKeyValuePairsWithOperator)
+    );
+  });
+
+  return filterList
+    .map(({ key, operator, value }) =>
+      formatFilter(key, operator, formatFilterValue(value))
+    )
+    .join(" AND ");
+}
+
+const getFiltersFromKeyValuePairs = (
+  keyValuePairs: [string, string | number | boolean][]
+): Filter[] => {
+  const filters: Filter[] = [];
+  keyValuePairs.forEach(([key, value]) => {
+    filters.push({
+      key,
+      operator: ":",
+      value: value,
+    });
+  });
+
+  return filters;
+};
+
+const getFiltersFromKeyValuePairsWithOperator = (
+  keyValuePairsWithOperator: [string, { startsWith: string }][]
+): Filter[] => {
+  const filters: Filter[] = [];
+  keyValuePairsWithOperator.forEach(([key, value]) => {
+    if ("startsWith" in value && typeof value.startsWith === "string") {
+      filters.push({
+        key,
+        operator: "^",
+        value: value.startsWith,
+      });
+    }
+  });
+
+  return filters;
+};
+
+const isSimpleValue = (value: unknown) => {
+  if (
+    typeof value === "string" ||
+    typeof value === "number" ||
+    typeof value === "boolean"
+  ) {
+    return true;
+  }
+  return false;
+};
+
+const isValueWithOperator = (value: unknown) => {
+  if (typeof value === "object" && value !== null && "startsWith" in value) {
+    return true;
+  }
+  return false;
+};
+
+const formatFilter = (key: string, operator: ":" | "^", value: string) => {
+  return `${key}${operator}${value}`;
+};
+
+const formatFilterKey = (key: string, nestedKey?: string) => {
+  if (nestedKey) {
+    return `${key}["${nestedKey}"]`;
```

Comment:
Be careful with trusting user inputted values (which you should assume these inputs are) when formatting. You should ensure you're properly escaping quotation marks to avoid "insertion attacks" here, or validate upfront that certain characters won't exist in the key/nestedKey parts.

Probably best to validate that keys don't include specific chars, and to allow any char in the values, but always escape them.

For example, currently it would be possible to do this at the key level:

```ts
{
  index: { "evil\"]:3 index[\"qux": "bar" },
}
```

Which will produce the query: `index["evil"]:3 index["qux"]:"bar"`.

Similarly, but at the value level:

```ts
{
  "org": "my-org\" OR org:\"another-org"
}
```

…which would produce:

`org:"my-org" OR org:"another-org"`

Reply:
Would escaping `"` quotes be enough? 

We allow anything but `-` and `+` as the starting characters for metadata keys as seen [here](https://github.com/liveblocks/liveblocks-backend/blob/c9e58d74ef0de67bc785af76f08b355ce9541988/rest-api/src/libs/decoders/index.ts#L37).

Reply:
Yeah, it would. The easiest way to do that is to use `JSON.stringify()`, which I'm [also using here](https://github.com/liveblocks/liveblocks-cloudflare/blob/main/packages/liveblocks-query-parser/src/prettify/index.ts#L13C38-L13C65).

Reply:
@nvie Thanks for all your feedback, I've adjusted the PR, should be good to go now!